modules/340-monitoring-ping/images/monitoring-ping/src/pkg/fastping/sockets.go

```diff
@@ -0,0 +1,360 @@
+// Package ping Copyright 2025 Flant JSC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fastping
+
+import (
+	"bytes"
+	"context"
+	"encoding/binary"
+	"fmt"
+	"net"
+	"sync"
+	"syscall"
+	"time"
+
+	"github.com/deckhouse/deckhouse/pkg/log"
+)
+
+type pendingPacket struct {
+	host     string
+	sentTime int64
+}
+
+type socketConn struct {
+	fd          int // raw socket file descriptor
+	id          int
+	seqPerHost  map[string]int
+	pending     sync.Map // thread-safe map: key -> pendingPacket
+	pendingLock sync.Mutex
+}
+
+func newSocket(ctx context.Context) (*socketConn, error) {
+	// Create raw ICMP socket
+	fd, err := syscall.Socket(syscall.AF_INET, syscall.SOCK_RAW, syscall.IPPROTO_ICMP)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create raw socket: %w", err)
+	}
+
+	// Set socket to non-blocking mode
+	err = syscall.SetNonblock(fd, true)
+	if err != nil {
+		syscall.Close(fd)
+		return nil, fmt.Errorf("failed to set non-blocking mode: %w", err)
+	}
```

Comment:
what if we check the error in a defer call and close the socket if the error isn't nil?

```suggestion
	fd, socketErr := syscall.Socket(syscall.AF_INET, syscall.SOCK_RAW, syscall.IPPROTO_ICMP)
	if socketErr != nil {
		return nil, fmt.Errorf("failed to create raw socket: %w", socketErr)
	}
	
	defer func() {
	        if socketErr != nil {
	                syscall.Close(fd)
	        }
	}()

	// Set socket to non-blocking mode
	socketErr = syscall.SetNonblock(fd, true)
	if socketErr != nil {
		return nil, fmt.Errorf("failed to set non-blocking mode: %w", socketErr)
	}
```

Reply:
Thanks for the suggestion! I tried using a defer with a shared error var, but I found it a bit harder to follow. Explicit close-on-error at each step seems clearer and more in line with Go’s straightforward style.